backup/backup.go

```diff
@@ -0,0 +1,123 @@
+package backup
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	help "github.com/featureform/helpers"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"io/ioutil"
+	"time"
+)
+
+type ProviderType string
+
+const (
+	AZURE ProviderType = "AZURE"
+	LOCAL              = "LOCAL"
+)
+
+// Client Allows ETCD to be tested with mock values
+type Client interface {
+	Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error)
+}
+
+type Backup struct {
+	ETCDClient   Client
+	ProviderType ProviderType
+}
+
+func (b *Backup) Save(filename string) error {
+	err := b.takeSnapshot(filename)
+	if err != nil {
+		return fmt.Errorf("could not take snapshot: %v", err)
+	}
+
+	backupProvider, err := b.getBackupProvider()
+	if err != nil {
+		return err
+	}
+	err = backupProvider.Init()
+	if err != nil {
+		return fmt.Errorf("could not initialize provider: %v", err)
+	}
+	err = backupProvider.Upload(filename, filename)
+	if err != nil {
+		return fmt.Errorf("cannot upload snapshot to filestore: %v", err)
+	}
+	return nil
+}
+
+func (b *Backup) Restore() error {
+	return nil
+}
+
+func (b *Backup) getBackupProvider() (Provider, error) {
+	var backupProvider Provider
+	switch b.ProviderType {
+	case AZURE:
+		backupProvider = &Azure{
+			AzureStorageAccount: help.GetEnv("AZURE_STORAGE_ACCOUNT", ""),
```

Comment:
I don't like using env variables when its not easy to override them in the code. It almost always is the wrong form and makes testing weird.

Reply:
Move this code into the Init methods